dask/distributed/worker.py

```diff
@@ -371,6 +375,8 @@ def block(self):
         """
         self._listen_workers_thread.join()
         self._listen_scheduler_thread.join()
+        self._heartbeat_thread.event.set()  # stop heartbeat
```

Comment:
You probably want to set the event right after the `self.status = 'closed'` line